# Chapter 1: SMPC Example



| **Platform**           | **Link**                                                                                                                                                                                                                                                                                               | **Notes**                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
|------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| _Local Folder_         | [Deep Dive Example: SMPC Example](1_privacy/Chapter_1_SMPC_Example.ipynb)                                                                                                                                                                                                                                |    `1_privacy/`<br>`Chapter_1_SMPC_Example.ipynb`<br>Recommend running locally                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
| _Book Chapter_         | [Chapter 1: Privacy](https://learning.oreilly.com/library/view/practicing-trustworthy-machine/9781098120269/ch01.html)                                                                                                                                                                                 | Ebook available from [O'Reilly](https://www.oreilly.com/library/view/practicing-trustworthy-machine/9781098120269/).<br>Physical book available from [Amazon](https://www.amazon.com/Practicing-Trustworthy-Machine-Learning-Transparent/dp/1098120272)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
| _GitHub_               | [![GitHub](https://img.shields.io/badge/-View%20on%20GitHub-181717?logo=github&logoColor=ffffff)](https://github.com/matthew-mcateer/practicing_trustworthy_machine_learning/blob/main/1_privacy/Chapter_1_SMPC_Example.ipynb)                                                                          | Please star on GitHub so others can<br>find this repo and get help with implenting<br>their trustworthy AI pipelines<br> [![GitHub forks](https://img.shields.io/github/forks/matthew-mcateer/practicing_trustworthy_machine_learning.svg?style=social&label=Fork&maxAge=2592000)](https://github.com/matthew-mcateer/practicing_trustworthy_machine_learning/network/) <br> [![GitHub stars](https://img.shields.io/github/stars/matthew-mcateer/practicing_trustworthy_machine_learning.svg?style=social&label=Star&maxAge=2592000)](https://github.com/matthew-mcateer/practicing_trustworthy_machine_learning/stargazers/) <br> [![GitHub watchers](https://img.shields.io/github/watchers/matthew-mcateer/practicing_trustworthy_machine_learning.svg?style=social&label=Watch&maxAge=2592000)](https://github.com/matthew-mcateer/practicing_trustworthy_machine_learning/watchers/) <br>[`@matthew-mcateer`](https://github.com/matthew-mcateer) [![GitHub followers](https://img.shields.io/github/followers/matthew-mcateer?style=social&label=Follow&maxAge=2592000)](https://github.com/matthew-mcateer?tab=followers)<br>[`@pruksmhc`](https://github.com/pruksmhc) [![GitHub followers](https://img.shields.io/github/followers/pruksmhc?style=social&label=Follow&maxAge=2592000)](https://github.com/pruksmhc?tab=followers)<br>[`@shubhobm`](https://github.com/shubhobm) [![GitHub followers](https://img.shields.io/github/followers/shubhobm?style=social&label=Follow&maxAge=2592000)](https://github.com/shubhobm?tab=followers)  |
| _Colab_                | [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/matthew-mcateer/practicing_trustworthy_machine_learning/blob/main/1_privacy/Chapter_1_SMPC_Example.ipynb)                                                                         | [Hacks required](https://gist.github.com/mwufi/6718b30761cd109f9aff04c5144eb885) to<br>enable docker support<br>Enable GPU<br>(Colab Pro recommended)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| _Kaggle_               | [ ![ Kaggle ]( https://kaggle.com/static/images/open-in-kaggle.svg ) ]( https://kaggle.com/kernels/welcome?src=https://github.com/matthew-mcateer/practicing_trustworthy_machine_learning/blob/main/1_privacy/Chapter_1_SMPC_Example.ipynb )                                                            | Not recommended due<br>to lack of docker support                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |
| _Gradient_             | [ ![ Gradient ]( https://assets.paperspace.io/img/gradient-badge.svg ) ]( https://console.paperspace.com/github/matthew-mcateer/practicing_trustworthy_machine_learning/blob/main/1_privacy/Chapter_1_SMPC_Example.ipynb )                                                                              | Enable GPU                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
| _SageMaker Studio Lab_ | [![ Open In SageMaker Studio Lab ]( https://studiolab.sagemaker.aws/studiolab.svg ) ]( https://studiolab.sagemaker.aws/import/github/matthew-mcateer/practicing_trustworthy_machine_learning/blob/main/1_privacy/Chapter_1_SMPC_Example.ipynb )                                                         | Recommend `ml.g4dn.xlarge` instance at minimum<br>(alternatively, `g4dn.2xlarge` if using EC2 instance)<br>See [Sagemaker docs](https://docs.aws.amazon.com/sagemaker/latest/dg/docker-containers.html)<br>regarding docker support                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |
| _Binder_               | [ ![ Binder ]( https://mybinder.org/badge_logo.svg ) ]( https://mybinder.org/v2/gh/matthew-mcateer/practicing_trustworthy_machine_learning/HEAD?urlpath=https%3A%2F%2Fgithub.com%2Fmatthew-mcateer%2Fpracticing_trustworthy_machine_learning%2Fblob%2Fmain%2F1_privacy%2FChapter_1_SMPC_Example.ipynb ) | Not recommended due to<br>lack of Docker support                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |


<!--
Originally found on GitHub at https://github.com/matthew-mcateer/practicing_trustworthy_machine_learning/blob/main/1_privacy/Chapter_1_SMPC_Example.ipynb
-->


The following is heavily based on OpenMined tutorials written by [Hrishikesh Kamath](https://twitter.com/kamathhrishi), [George Muraru](https://twitter.com/GeorgeMuraru), and [Théo Ryffel](http://twitter.com/theoryffel)


This tutorial has seen the most changes out of any in the book. **The code for the early release version of the book [can be found in the `deprecated` folder](https://github.com/matthew-mcateer/practicing_trustworthy_machine_learning/blob/main/1_privacy/deprecated/Chapter_1_SMPC_Example_DEPRECATED.ipynb), and is conceptually identical to the tutorial here, but thanks to the `SyMPC` package this version provides a much gentler introduct to steps such as uploading one's data before any computation takes place.** The original code is still valid with the package versions that have been specified, but Google Colab no longer supports those specific tutorials.

**UPDATE:** The `SyMPC` package no longer functions due to the removal of the `sympc-dev` branch of PySyft. As such, the dependencies for this tutorial are no longer resolvable. The dependencies for the specific versions of the packages used in the original tutorial are also no longer resolvable. We will provide an updated version of this tutorial in the future, though all the concepts described still apply.

## ⚙️ Setup

In [1]:
#@title Checking environment
%load_ext watermark
%watermark -a "Practicing Trustworthy machine Learning" -u -d -v -m -p syft,loguru,torch,sympc,pandas,numpy,crypten

Author: Practicing Trustworthy machine Learning

Last updated: 2022-11-23

Python implementation: CPython
Python version       : 3.8.16
IPython version      : 8.4.0

syft   : 0.3.0
loguru : 0.7.0
torch  : 1.7.1
sympc  : 0.5.0rc1.post0.dev92+g278a13d.dirty
pandas : 1.8.1
numpy  : 1.18.5
crypten: 0.4.1

Compiler    : GCC 11.3.0
OS          : Linux
Release     : 5.15.0-1036-aws
Machine     : x86_64
Processor   : x86_64
CPU cores   : 4
Architecture: 64bit


In [2]:
import pandas as pd
import numpy as np
import time
import torch

#Set a manual seed to maintain consistency
torch.manual_seed(1337)

## 🗂️ Data Loading and Processing for SMPC

Secure Multi-Party Computation (SMPC) is a subfield of cryptography that allows multiple parties to jointly compute a function over their inputs while keeping those inputs private. This is done by splitting the inputs into shares, and distributing them among the parties. The parties then perform computations on their shares, and share the results with each other. The parties can then combine the results to get the final result of the computation.

This is an almost insultingly simple overview of a general class of protocols. Many of these are complex, or are based on assumptions that only apply in key situations.


For the sake of our demonstration, we'll be training a regression network on the [Boston Housing Dataset](https://www.kaggle.com/datasets/altavish/boston-housing-dataset), and then running inference on the trained model using SMPC.

The Boston housing dataset contains 506 observations and 14 variables.

It also contains some missing values (which, if we are doing SMPC for real, we should expect to have to deal with varying levels of data quality).


In [3]:
# Download Boston housing dataset
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/housing/housing.data

--2022-11-23 21:31:00--  https://archive.ics.uci.edu/ml/machine-learning-databases/housing/housing.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 49082 (48K) [application/x-httpd-php]
Saving to: ‘housing.data’

housing.data        100%[===================>]  47.93K   157KB/s    in 0.3s    

2022-11-23 21:31:02 (157 KB/s) - ‘housing.data’ saved [49082/49082]



Once downloaded, we can then import the dataset in CSV format and add the data headers.

In [4]:
# Import dataset and add headers
dataset = pd.read_csv(
    "housing.data",
    delim_whitespace=True,
    names=[
        "crim",
        "zn",
        "indus",
        "chas",
        "nox",
        "rm",
        "age",
        "dis",
        "rad",
        "tax",
        "ptratio",
        "black",
        "lstat",
        "medv",
    ],
)

While we may not necessarily have the opportunity in a SMPC scenario (depending on the domain, this might be either impossible or strongly discouraged), we will look at our data to get a sense of what we're working with.

In [5]:
dataset.head()

crim    zn  indus  chas    nox     rm   age     dis  rad    tax  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   ptratio   black  lstat  medv  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

The data is looking good, though we should also make sure we're going through the proper preprocessing steps.

For starters, we want to make sure we're selecting the right column to predict by splitting data into the features and target variables.

In [6]:
features = dataset.drop("medv", axis=1)
targets = dataset["medv"]

For the predictive features, let's make sure we're normalizing the data to reduce the impact of outliers.

In [7]:
features = features.apply(
    lambda x: (x - x.mean()) / x.std()
)

Now, we should have the data features we want to input into our model...

In [8]:
features

crim        zn     indus      chas       nox        rm       age  \
0   -0.419367  0.284548 -1.286636 -0.272329 -0.144075  0.413263 -0.119895   
1   -0.416927 -0.487240 -0.592794 -0.272329 -0.739530  0.194082  0.366803   
2   -0.416929 -0.487240 -0.592794 -0.272329 -0.739530  1.281446 -0.265549   
3   -0.416338 -0.487240 -1.305586 -0.272329 -0.834458  1.015298 -0.809088   
4   -0.412074 -0.487240 -1.305586 -0.272329 -0.834458  1.227362 -0.510674   
..        ...       ...       ...       ...       ...       ...       ...   
501 -0.412820 -0.487240  0.115624 -0.272329  0.157968  0.438881  0.018654   
502 -0.414839 -0.487240  0.115624 -0.272329  0.157968 -0.234316  0.288648   
503 -0.413038 -0.487240  0.115624 -0.272329  0.157968  0.983986  0.796661   
504 -0.407361 -0.487240  0.115624 -0.272329  0.157968  0.724955  0.736268   
505 -0.414590 -0.487240  0.115624 -0.272329  0.157968 -0.362408  0.434302   

          dis       rad       tax   ptratio     black     lstat  
0    0.140075 -0.981871 -0.665949 -1.457558  0.440616 -1.074499  
1    0.556609 -0.867024 -0.986353 -0.302794  0.440616 -0.491953  
2    0.556609 -0.867024 -0.986353 -0.302794  0.396035 -1.207532  
3    1.076671 -0.752178 -1.105022  0.112920  0.415751 -1.360171  
4    1.076671 -0.752178 -1.105022  0.112920  0.440616 -1.025487  
..        ...       ...       ...       ...       ...       ...  
501 -0.625178 -0.981871 -0.802418  1.175303  0.386834 -0.417734  
502 -0.715931 -0.981871 -0.802418  1.175303  0.440616 -0.500355  
503 -0.772919 -0.981871 -0.802418  1.175303  0.440616 -0.982076  
504 -0.667776 -0.981871 -0.802418  1.175303  0.402826 -0.864446  
505 -0.612640 -0.981871 -0.802418  1.175303  0.440616 -0.668397  

[506 rows x 13 columns]

...as well as the targets we want to predict.

In [9]:
targets

0      24.0
1      21.6
2      34.7
3      33.4
4      36.2
       ... 
501    22.4
502    20.6
503    23.9
504    22.0
505    11.9
Name: medv, Length: 506, dtype: float64

Since we'll be using PyTorch, we should convert our data into tensors that PyTorch can use.

In [10]:
features = torch.tensor(features.values.astype(np.float32))
targets = torch.tensor(targets.values.astype(np.float32))

We'll define our aruments for training, such as the batch size, epochs, train/test split, and learning rate.

Normally we'd put more effort into creating callbacks that could optimize the relevant hyperparameters on the fly, but we're making the training phase as simple as possible to illustrate that the special sauce is coming at the inference instead.

In [11]:
batch_size = 16
epochs = 300
train_test_split = 0.8
lr = 0.001

We will split the dataset into the train and test slices.

In [12]:
train_indices = int(len(features) * train_test_split)

train_x = features[:train_indices]
train_y = targets[:train_indices]

test_x = features[train_indices + 1 :]
test_y = targets[train_indices + 1 :]

And we will divide the dataset into branches.

In [13]:
def get_batches(X, y):
    batches = []
    for index in range(0, len(train_x) + 1, batch_size):
        batches.append((X[index : index + batch_size], y[index : index + batch_size]))

    return batches

In [14]:
train_batches = get_batches(train_x, train_y)

## 🗝️ Plaintext Training

We will demonstrate SMPC as it pertains to running inference on an already-trained model.
We will use a toy linear regression model (see Chapter 3 for more on model explainability) on a toy housing dataset.

For starters, let's import the PySyft library.

In [15]:
import syft as sy
sy.logger.remove()

Next, we define a basic linear regression model.

This behaves just like any other `nn.Module` network we'd definine in PyTorch save for two key differences:

1. We replace the `nn.Module` with `sy.Module`.
2. We wrap up the reference to pytorch itself with the `torch_ref` argument.

In [16]:
class LinearSyNet(sy.Module):
    def __init__(self, torch_ref):
        super(LinearSyNet, self).__init__(torch_ref=torch_ref)
        self.fc1 = self.torch_ref.nn.Linear(13,1)

    def forward(self, x):
        x = self.fc1(x)
        return x


Next we initialize our `model`, along with the optimizer and loss function.

In [17]:
model = LinearSyNet(torch)
loss_function = torch.nn.MSELoss(reduction='mean') 
optimizer = torch.optim.SGD(model.parameters(), lr=lr)

And we can run our standard Stochastic Gradient Descent training loop.

For a dataset this simple we're not even really bothering with features like momentum.

In [18]:
for epoch in range(epochs):
    running_loss = 0.0
    for index in range(0, len(train_batches)):
        
        outputs = model(train_batches[index][0]).reshape([-1])
        loss = loss_function(outputs, train_batches[index][1])
        running_loss += loss
        loss.backward()
        optimizer.step()

    test_accuracy = loss_function(model(test_x).reshape([-1]), test_y)
    if (epoch % 50) == 0:
        print(
            (
                f"Epoch {epoch}/{epochs}  "
                f"Running Loss : {running_loss.item()/batch_size} and "
                f"test loss: {test_accuracy.item()}"
            )
        )

Epoch 0/300  Running Loss : 998.8542400885754 and test loss: 385.4494185937673
Epoch 50/300  Running Loss : 62.82259386625583 and test loss: 113.30346797719565
Epoch 100/300  Running Loss : 53.1354183670673 and test loss: 57.561960747373489
Epoch 150/300  Running Loss : 40.82360264512523 and test loss: 31.9676642938053
Epoch 200/300  Running Loss : 39.0269033569808 and test loss: 24.6837654675415
Epoch 250/300  Running Loss : 37.26858443477006 and test loss: 21.7789115590072


## 🔓 Plaintext Inference

Before we run encrypted inference on our model, we want to have a baseline for comparison.

If you're running SMPC in real-life, then chances are you're running it on a network.

Distributed systems engineering is very unforgiving when it comes to increasing the runtime of your code.

For this reason, we'll make sure to include an unencrypted inference baseline to compare against in the future.

In [19]:
start_time = time.time()
# Perform inference in plaintext
plaintext_predictions = model(test_x).reshape([-1])
end_time = time.time()

In [20]:
print("Plaintext (Local) prediction:")
print("MSE Loss: ", loss_function(plaintext_predictions, test_y).item())
print("Inference time: ", str(end_time - start_time), "s")

Plaintext (Local) prediction:
MSE Loss:  21.778203758424003
Inference time:  0.0007644155007102572 s


Now we can run inference over our secure multi-party computation network.

## 🔐 Encrypted Inference

This is where we will import the `sympc` library.
This library is a wrapper on top of PySyft that allows for SMPC.
More importantly, it contains code for a variety of SMPC protocols for varying levels of network sizes and network trust.

In [21]:
#SyMPC imports required for encrypted inference
import sympc
from sympc.session import Session
from sympc.session import SessionManager
from sympc.tensor import MPCTensor,ShareTensorPointer
from sympc.protocol import ABY3,beaver,Falcon,FSS,spdz

The `MPCTensor` class is the tensor that holds reference to the shares owned by the different parties.
The `Session` is a key object that stores many information about the current computation, including a reference to the parties involved.

We will create a utility function that spins up an arbitrary number of PySyft `VirtualMachine` workers.

In [22]:
def get_clients(n_parties: int) -> list[sy.VirtualMachine]:
    """Generate required number of syft clients and return them.
    """

    parties = []
    for index in range(n_parties):
        parties.append(
            sy.VirtualMachine(name="worker" + str(index)).get_root_client()
        )

    return parties

We will also make sure to effectively split up our data into a number of chunks equal to the number of parties in our SMPC network and distribute it appropriately.

We have different ways to do this and all of them deal with `MPCTensor`, an orchestator that wants to do computation on data it does not see.

In [23]:
def split_send(data: MPCTensor, session: Session) -> list[ShareTensorPointer]:
    """Splits data into number of chunks equal to number of
    parties and distributes it to respective parties.
    """
    data_pointers = []

    split_size = int(len(data) / len(session.parties)) + 1
    for index in range(0, len(session.parties)):
        ptr = data[index * split_size : index * split_size + split_size].share(
            session=session
        )
        data_pointers.append(ptr)

    return data_pointers


We will then create our function for inference.
This is the crucial component as which protocol and which settings we use will heavily depend on

1. The mathematical operations we want to perform.
2. The number of parties involved.
3. The trust level between the parties (and thus the security level we need).

SyMPC supports several protocols that one might find handy in different situations.

- `Falcon`
    - Our main protocol of interest, due to its ability to be tuned depending on the level of trust in the network. Falcon uses 2-out-of-3 sharing where each party recieves two shares allowing 2 parties to reconstruct a secret without other party knowing. Thus, it is fast and secure for 3 participants, but it's not really scalable beyond 3. Falcon also assumes that majority of parties are honest (2 in this case).
    - Paper: [FALCON: Honest-Majority Maliciously Secure Framework for Private Deep Learning](https://arxiv.org/abs/2004.02229) 
    - OpenMined/SyMPC Implementation: [code](https://github.com/OpenMined/SyMPC/blob/main/src/sympc/protocol/falcon/falcon.py)
- `ABY3`:
    - Another MPC framework for Machine learning.
    - Paper: [ABY3 : A Mixed Protocol Framework for Machine Learning](https://eprint.iacr.org/2018/403.pdf) 
    - OpenMined/SyMPC Implementation: [code](https://github.com/OpenMined/SyMPC/blob/main/src/sympc/protocol/aby3/aby3.py)
- `Function Secret Sharing (FSS)`:
    - Securely works for MPC involving two parties. Distributes a single share of the data to both parties, and requires both parties to reconstruct the data.
    - Paper: [ARIANN: Low-Interaction Privacy-Preserving Deep Learning via Function Secret Sharing](https://arxiv.org/abs/2006.04593) 
    - OpenMined/SyMPC Implementation: [code](https://github.com/OpenMined/SyMPC/blob/main/src/sympc/protocol/fss/fss.py)
- `SPDZ`:
    - Pronounced "speedz", this protocol can be extended to an arbitrary number of parties in an SMPC network, and is secure against an active adversary corrupting up to N-1 of the N participants. SPDZ and FSS distributes a single share to every party requiring shares from all parties for reconstruction ensuring no parties could collude. Currently SyMPC provides support for SPDZ and FSS with semi-honest security guarantee. This allows parties to tamper with shares leading to incorrect results.
    - Paper: [Multiparty Computation from Somewhat Homomorphic Encryption](https://link.springer.com/chapter/10.1007/978-3-642-32009-5_38) 
    - OpenMined/SyMPC Implementation: [code]()
- `Beaver Triples`:
    - One of the earliest protocols for SMPC (created back in 1992).
    - Paper: [Efficient multiparty protocols using circuit randomization](https://link.springer.com/content/pdf/10.1007/3-540-46766-1_34.pdf) 
    - OpenMined/SyMPC Implementation: [code](https://github.com/OpenMined/SyMPC/blob/main/src/sympc/protocol/beaver/beaver.py)



In [24]:
def inference(n_clients: int, protocol=None) -> torch.Tensor:
    """Perform encrypted inference on the test dataset over the network.
    """
    # Get VM clients
    parties = get_clients(n_clients)

    # Setup the session for the computation
    if protocol:
        session = Session(parties=parties, protocol=protocol)
    else:
        session = Session(parties=parties)

    SessionManager.setup_mpc(session)

    # Split data and send data to clients
    pointers = split_send(test_x, session)

    # Encrypt model
    mpc_model = model.share(session)

    # Perform inference and measure time taken
    start_time = time.time()

    results = []

    for ptr in pointers:
        encrypted_results = mpc_model(ptr)
        plaintext_results = encrypted_results.reconstruct()
        results.append(plaintext_results)

    end_time = time.time()

    print(f"Time for inference: {end_time-start_time}s")

    predictions = torch.cat(results).reshape([-1])

    #Calculate Loss
    print("MSE Loss: ", loss_function(predictions,test_y).item())
    return predictions

Now we can run our inference on an example network.

### Encrypted Inference (assuming "semi-honest" nodes)

For the sake of example, lets' assume we have a network of 3 workers.

Let's also assume that our `Falcon` protocol is assuming "semi-honest" nodes.

In [25]:
print("Falcon on N=3 semi-honest workers:")
predictions = inference(
    n_clients=3,
    Falcon("semi-honest")
)

Falcon on N=3 semi-honest workers:
Time for inference: 0.1194368886979717s
MSE Loss:  21.778549300512648


We can then compare the results of our encrypted inference to our plaintext inference.

What do we find? The prediction values and mean squared error values are almost identical between the encrypted and plaintext versions.
There are some very minor differences in the outputs due to precision loss.
But, aside from that they perform indistinguishably well on the task at hand (or indistinguishably poorly, depending on how you look at it).

In [26]:
for index in range(0,10):
    print(f"Index {index}")
    print(f"Expected Prediction: {test_y[index]}")
    print(f"Plaintext Prediction Output {plaintext_predictions[index].item()}")
    print(f"Encrypted Prediction Output {predictions[index].item()}")
    print("\n")

Index 0
Expected Prediction: 5.0
Plaintext Prediction Output 2.862741785703401
Encrypted Prediction Output 2.8629437149153


Index 1
Expected Prediction: 11.899999618530273
Plaintext Prediction Output 4.893105021494213
Encrypted Prediction Output 4.893270911008


Index 2
Expected Prediction: 27.899999618530273
Plaintext Prediction Output 20.178911721169953
Encrypted Prediction Output 20.178953017991623


Index 3
Expected Prediction: 17.200000762939453
Plaintext Prediction Output 11.998541888269916
Encrypted Prediction Output 11.998668187754756


Index 4
Expected Prediction: 27.5
Plaintext Prediction Output 19.139752971707028
Encrypted Prediction Output 19.139784578666693

Index 5
Expected Prediction: 15.0
Plaintext Prediction Output 10.023104710183637
Encrypted Prediction Output 10.02325823903034


Index 6
Expected Prediction: 17.200000762939453
Plaintext Prediction Output 15.837065783018554
Encrypted Prediction Output 15.837146619743531


Index 7
Expected Prediction: 17.89999961853027

### Encrypted Inference (assuming "malicious" nodes)

The Falcon protocol can also be secured against "malicious" nodes hiding in an otherwise honest majority network.

What does this mean in practice?
It means that setting the security level to `malicious` will ensure that worker nodes avoid deviating from the protocol or tampering with the shares of data they're given.

In [27]:
print("Falcon on N=3 malicious workers:")
predictions = inference(
    n_clients=3,
    Falcon("malicious")
)

Falcon on N=3 malicious workers:
Time for inference: 0.647966524686441s
MSE Loss:  21.778416803181585


### SPDZ, and expanding the number of worker nodes

As mentioned when we set up our `inference()` function, SyMPC by default uses SPDZ and FSS protocol with "semi-honest" security type.

The SPDZ protocol can be used for multiplication and related operations (convolution,matmul,etc).

Functional Secret Sharing (FSS) for other operations such as comparison, equality, maxpool, etc.

The SMPC protocol goes beyond the types of operations.

FSS really only securely works for 2 parties.

On the other hand, SPDZ can be extended to `N` arbitrary parties.

Since linear regression uses only matmul, we can utilize the SPDZ protocol, which thus allows us to run linear regression with several parties in this tutorial.

In [28]:
print("SPDZ on N=3 semi-honest workers:")
predictions=inference(
    n_clients=3
)
print("\n")
print("SPDZ on N=5 semi-honest workers:")
predictions=inference(
    n_clients=5
)

SPDZ on N=3 semi-honest workers:
Time for inference: 0.4948297386359114s
MSE Loss:  21.778019022766458

SPDZ on N=5 semi-honest workers:
Time for inference: 1.31117314340541s
MSE Loss:  21.778948081205161


### Comparison

| **Protocol** | **Security Type** | **Parties** | **Inference Time (s)** |
|--------------|-------------------|-------------|------------------------|
| Plaintext    | N/A               | 1           | 0.00076441550071       |
| Falcon       | Semi-honest       | 3           | 0.11943688869797       |
| Falcon       | Malicious         | 3           | 0.64796652468644       |
| SPDZ         | Semi-honest       | 3           | 0.49482973863591       |
| SPDZ         | Semi-honest       | 5           | 1.31117314340541       |

This is only a very loose comparison, and your results may vary depending on your CPU hardware, or even if you're actually running the protocols in question on a network.

The Falcoln protocol is the fastest in a 3-node setting, while SPDZ allows for inference with greater than 3 parties.
Both of these protocols allow for inference with almost the same accuracy as plaintext.

## 📚 References, Resources, and Further Reading

This is obviously a very simplified approach to running secure multi-party computation.
Still, this is a good primer on the important foundational concepts.

For further resources on implementing SCMP, we recommend the following resources (some of which inspired the work in this tutorial):

- [OF PRIVACY AND PURSE STRINGS Subscribe Posted on January 8th, 2022 under Differential Privacy](https://blog.openmined.org/of-privacy-and-purse-strings/)
- [SyMPC/examples/Encrypted-Inference-LinearRegression.ipynb](https://github.com/OpenMined/SyMPC/blob/main/examples/Encrypted-Inference-LinearRegression.ipynb)

**OpenMined Syft Docs**
- [Installation guides](https://openmined.github.io/PySyft/getting_started/index.html):
    - [Linux](https://openmined.github.io/PySyft/install_tutorials/linux.html##)
    - [OSX](https://openmined.github.io/PySyft/install_tutorials/osx_11_5_1.html#)
    - [Windows](https://openmined.github.io/PySyft/install_tutorials/windows.html)
- [User guide index](https://openmined.github.io/PySyft/user_guide/index.html)
- [API Reference Index](https://openmined.github.io/PySyft/api_reference/index.html)
- [Glossary](https://openmined.github.io/PySyft/deployment/glossary.html#privacy-budget)

**OpenMined GitHub Repos**
- [OpenMined/PySyft](https://github.com/OpenMined/PySyft) - The main one we use in this tutorial
- [OpenMined/SyMPC](https://github.com/OpenMined/SyMPC) - A SMPC companion library for Syft
- [OpenMined/SyferText](https://github.com/OpenMined/SyferText) - A privacy preserving NLP framework
- [OpenMined/PyDP](https://github.com/OpenMined/PyDP) - The Python Differential Privacy Library. Built on top of: https://github.com/google/differential-privacy
- [OpenMined/PipelineDP](https://github.com/OpenMined/PipelineDP) - PipelineDP is a Python framework for applying differentially private aggregations to large datasets using batch processing systems such as Apache Spark, Apache Beam, and more.
- [OpenMined/TenSEAL](https://github.com/OpenMined/TenSEAL) - A library for doing homomorphic encryption operations on tensors
- [OpenMined/datasets](https://github.com/OpenMined/datasets) - Example datasets used in OpenMined Demos.